File: router.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

type FeeStructure struct {
	Maker float64
	Taker float64
	Note  string
}

type LiquidityData struct {
	Depth1Pct float64
	Depth2Pct float64
	Spread    float64
}

type PlatformResult struct {
	Name            string
	Price           float64
	Liquidity       float64
	Slippage        float64
	Fees            float64
	NetCost         float64
	FillProbability float64
	BalancedScore   float64
}

type Route struct {
	Platform         string  `json:"platform"`
	Mode             string  `json:"mode"`
	ExpectedPrice    float64 `json:"expectedPrice"`
	ExpectedSlippage float64 `json:"expectedSlippage"`
	Fees             float64 `json:"fees"`
	NetCost          float64 `json:"netCost"`
	FillProbability  float64 `json:"fillProbability"`
	Score            float64 `json:"score"`
}

type OrderExecution struct {
	OrderID        string    `json:"orderId"`
	Platform       string    `json:"platform"`
	FillPrice      float64   `json:"fillPrice"`
	ActualSlippage float64   `json:"actualSlippage"`
	Fees           float64   `json:"fees"`
	ExecutedAt     time.Time `json:"executedAt"`
}

type SplitLeg struct {
	Platform          string  `json:"platform"`
	Size              float64 `json:"size"`
	Price             float64 `json:"price"`
	EstimatedSlippage float64 `json:"estimatedSlippage"`
}

type SplitOrder struct {
	Legs          []SplitLeg `json:"legs"`
	TotalSlippage float64    `json:"totalSlippage"`
	AvgPrice      float64    `json:"avgPrice"`
}

type SmartRouter struct {
	platforms     []string
	defaultMode   string
	fees          map[string]FeeStructure
	liquidityData map[string]LiquidityData
	// marketIndex provides live prices from real prediction market APIs.
	marketIndex *MarketIndex
}

func NewSmartRouter(platforms []string, defaultMode string, mi *MarketIndex) *SmartRouter {
	if len(platforms) == 0 {
		platforms = []string{"polymarket", "kalshi", "manifold"}
	}
	if defaultMode == "" {
		defaultMode = "balanced"
	}
	return &SmartRouter{
		platforms:   platforms,
		defaultMode: defaultMode,
		marketIndex: mi,
		fees: map[string]FeeStructure{
			"polymarket": {Maker: 0, Taker: 0, Note: "Zero fees on most markets"},
			"kalshi":     {Maker: 0.17, Taker: 1.2, Note: "Formula-based, capped ~2%"},
			"manifold":   {Maker: 0, Taker: 0, Note: "Play money"},
		},
		liquidityData: map[string]LiquidityData{
			"polymarket": {Depth1Pct: 500000, Depth2Pct: 1000000, Spread: 0.02},
			"kalshi":     {Depth1Pct: 250000, Depth2Pct: 600000, Spread: 0.05},
			"manifold":   {Depth1Pct: 10000, Depth2Pct: 50000, Spread: 0.1},
		},
	}
}

func (r *SmartRouter) FindBestRoute(market, side string, size float64, mode string) (*Route, error) {
	if mode == "" {
		mode = r.defaultMode
	}
	comparison := r.compare(market, side, size)

	switch mode {
	case "best-price":
		best := comparison[0]
		for _, p := range comparison[1:] {
			if p.Price < best.Price {
				best = p
			}
		}
		return r.createRoute(best, mode), nil

	case "best-liquidity":
		best := comparison[0]
		for _, p := range comparison[1:] {
			if p.Liquidity > best.Liquidity {
				best = p
			}
		}
		return r.createRoute(best, mode), nil

	case "lowest-fee":
		best := comparison[0]
		for _, p := range comparison[1:] {
			if p.Fees < best.Fees {
				best = p
			}
		}
		return r.createRoute(best, mode), nil

	default:
		w := struct{ Price, Liquidity, Fees float64 }{0.4, 0.3, 0.3}

		bestPrice := comparison[0].Price
		bestLiquidity := comparison[0].Liquidity
		lowestFees := comparison[0].Fees
		for _, p := range comparison[1:] {
			if p.Price < bestPrice {
				bestPrice = p.Price
			}
			if p.Liquidity > bestLiquidity {
				bestLiquidity = p.Liquidity
			}
			if p.Fees < lowestFees {
				lowestFees = p.Fees
			}
		}

		for i := range comparison {
			priceScore := bestPrice / comparison[i].Price
			liquidityScore := comparison[i].Liquidity / bestLiquidity
			feeScore := lowestFees / (comparison[i].Fees + 0.01)
			comparison[i].BalancedScore = priceScore*w.Price + liquidityScore*w.Liquidity + feeScore*w.Fees
		}

		best := comparison[0]
		for _, p := range comparison[1:] {
			if p.BalancedScore > best.BalancedScore {
				best = p
			}
		}
		return r.createRoute(best, mode), nil
	}
}

func (r *SmartRouter) compare(market, side string, size float64) []PlatformResult {
	// Pull the live base price from the MarketIndex (real API data).
	basePrice := r.getLivePrice(market)

	results := make([]PlatformResult, 0, len(r.platforms))
	for _, platform := range r.platforms {
		price := r.getPlatformPrice(platform, basePrice, side)

		feeStructure := r.fees[platform]
		fees := (size * feeStructure.Taker) / 100

		liq := r.liquidityData[platform]
		var slippage float64
		if size > liq.Depth1Pct {
			slippage = 2
		} else {
			slippage = (size / liq.Depth1Pct) * 100
		}

		results = append(results, PlatformResult{
			Name:            platform,
			Price:           price,
			Liquidity:       liq.Depth1Pct,
			Slippage:        slippage,
			Fees:            fees,
			NetCost:         size*price + fees,
			FillProbability: math.Max(0, 100-slippage),
		})
	}
	return results
}

func (r *SmartRouter) Execute(route *Route) *OrderExecution {
	spread := (rand.Float64() - 0.5) * route.ExpectedSlippage / 100
	actualPrice := route.ExpectedPrice * (1 + spread)
	actualSlippage := math.Abs((actualPrice-route.ExpectedPrice)/route.ExpectedPrice) * 100

	return &OrderExecution{
		OrderID:        fmt.Sprintf("ord_%d_%s", time.Now().UnixMilli(), randStr(8)),
		Platform:       route.Platform,
		FillPrice:      actualPrice,
		ActualSlippage: actualSlippage,
		Fees:           route.Fees,
		ExecutedAt:     time.Now(),
	}
}

func (r *SmartRouter) SplitOrder(market, side string, size, maxSlippage float64) *SplitOrder {
	if maxSlippage == 0 {
		maxSlippage = 0.02
	}
	comparison := r.compare(market, side, size)

	viable := []PlatformResult{}
	for _, p := range comparison {
		if p.Slippage <= maxSlippage {
			viable = append(viable, p)
		}
	}

	if len(viable) == 0 {
		return &SplitOrder{Legs: []SplitLeg{}, TotalSlippage: maxSlippage + 1}
	}

	totalLiquidity := 0.0
	for _, p := range viable {
		totalLiquidity += p.Liquidity
	}

	legs := []SplitLeg{}
	remainingSize := size

	for i := 0; i < len(viable)-1; i++ {
		p := viable[i]
		allocation := math.Min(p.Liquidity*0.8, remainingSize*(p.Liquidity/totalLiquidity))
		legs = append(legs, SplitLeg{Platform: p.Name, Size: allocation, Price: p.Price, EstimatedSlippage: p.Slippage})
		remainingSize -= allocation
	}

	last := viable[len(viable)-1]
	legs = append(legs, SplitLeg{Platform: last.Name, Size: remainingSize, Price: last.Price, EstimatedSlippage: last.Slippage})

	avgPrice := 0.0
	totalSlippage := 0.0
	for _, leg := range legs {
		avgPrice += leg.Price
		totalSlippage += leg.EstimatedSlippage
	}
	avgPrice /= float64(len(legs))
	totalSlippage /= float64(len(legs))

	return &SplitOrder{Legs: legs, TotalSlippage: totalSlippage, AvgPrice: avgPrice}
}

// getLivePrice fetches the current YES probability for a market from the live
// MarketIndex (which pulls from real Manifold / Polymarket APIs).
func (r *SmartRouter) getLivePrice(market string) float64 {
	if r.marketIndex != nil {
		return r.marketIndex.GetPrice(market)
	}
	return 0.50
}

// getPlatformPrice applies a small per-platform spread on top of the live base price.
// Real arbitrage differences between platforms are typically < 1–2%.
func (r *SmartRouter) getPlatformPrice(platform string, basePrice float64, side string) float64 {
	spreads := map[string]float64{
		"polymarket": 0.000, // tightest spread, highest liquidity
		"kalshi":     0.010, // ~1% wider
		"manifold":   0.020, // play money, wider spreads
	}
	spread, ok := spreads[platform]
	if !ok {
		spread = 0.005
	}
	// For YES bets, ask price is base + half spread; for NO it's (1-base) + half spread
	if side == "NO" {
		basePrice = 1.0 - basePrice
	}
	return math.Min(0.99, math.Max(0.01, basePrice+spread))
}

func (r *SmartRouter) createRoute(p PlatformResult, mode string) *Route {
	return &Route{
		Platform:         p.Name,
		Mode:             mode,
		ExpectedPrice:    p.Price,
		ExpectedSlippage: p.Slippage,
		Fees:             p.Fees,
		NetCost:          p.NetCost,
		FillProbability:  p.FillProbability,
		Score:            p.BalancedScore,
	}
}

File: src/router.js
/**
 * Smart Router - Routes orders to best platform based on price, liquidity, fees
 * Based on CloddsBot router skill documentation
 * Source: tessl.io/registry/skills/github/alsk1992/CloddsBot/router
 */

class SmartRouter {
  constructor(options = {}) {
    this.platforms = options.platforms || ['polymarket', 'kalshi', 'manifold'];
    this.defaultMode = options.defaultMode || 'balanced';
    
    // Fee structures (based on documentation)
    this.fees = {
      polymarket: { maker: 0, taker: 0, note: 'Zero fees on most markets' },
      kalshi: { maker: 0.17, taker: 1.2, note: 'Formula-based, capped ~2%' },
      manifold: { maker: 0, taker: 0, note: 'Play money' }
    };
    
    // Mock liquidity data (in production, fetch from API)
    this.liquidityData = {
      polymarket: { depth1Pct: 500000, depth2Pct: 1000000, spread: 0.02 },
      kalshi: { depth1Pct: 250000, depth2Pct: 600000, spread: 0.05 },
      manifold: { depth1Pct: 10000, depth2Pct: 50000, spread: 0.1 }
    };
  }

  /**
   * Find best route for an order
   * @param {Object} params - Order parameters
   * @param {string} params.market - Market identifier
   * @param {string} params.side - 'YES'/'NO' or 'long'/'short' or 'buy'/'sell'
   * @param {number} params.size - Order size in dollars
   * @param {string} params.mode - 'best-price' | 'best-liquidity' | 'lowest-fee' | 'balanced'
   * @param {Object} params.weights - Custom weights for balanced mode
   */
  async findBestRoute({ market, side, size, mode = this.defaultMode, weights = null }) {
    const comparison = await this.compare({ market, side, size });
    
    if (mode === 'best-price') {
      const best = comparison.reduce((best, curr) => 
        curr.price < best.price ? curr : best
      );
      return this.createRoute(best, mode);
    }
    
    if (mode === 'best-liquidity') {
      const best = comparison.reduce((best, curr) => 
        curr.liquidity > best.liquidity ? curr : best
      );
      return this.createRoute(best, mode);
    }
    
    if (mode === 'lowest-fee') {
      const best = comparison.reduce((best, curr) => 
        curr.fees < best.fees ? curr : best
      );
      return this.createRoute(best, mode);
    }
    
    // Balanced mode
    const w = weights || { price: 0.4, liquidity: 0.3, fees: 0.3 };
    
    for (const platform of comparison) {
      // Normalize scores
      const bestPrice = Math.min(...comparison.map(p => p.price));
      const bestLiquidity = Math.max(...comparison.map(p => p.liquidity));
      const lowestFees = Math.min(...comparison.map(p => p.fees));
      
      const priceScore = bestPrice / platform.price;
      const liquidityScore = platform.liquidity / bestLiquidity;
      const feeScore = lowestFees / (platform.fees + 0.01);
      
      platform.balancedScore = 
        (priceScore * w.price) +
        (liquidityScore * w.liquidity) +
        (feeScore * w.fees);
    }
    
    const best = comparison.reduce((best, curr) => 
      curr.balancedScore > best.balancedScore ? curr : best
    );
    
    return this.createRoute(best, mode);
  }

  /**
   * Compare all platforms for an order
   * @param {Object} params - Order parameters
   */
  async compare({ market, side, size }) {
    const results = [];
    
    for (const platform of this.platforms) {
      // Mock price data (in production, fetch from market API)
      const basePrice = this.getBasePrice(market);
      const price = this.getPlatformPrice(platform, basePrice, side);
      
      // Calculate fees
      const feeStructure = this.fees[platform];
      const fees = (size * feeStructure.taker) / 100;
      
      // Get liquidity data
      const liquidity = this.liquidityData[platform];
      const slippage = size > liquidity.depth1Pct ? 2 : (size / liquidity.depth1Pct) * 100;
      
      results.push({
        name: platform,
        price: price,
        liquidity: liquidity.depth1Pct,
        slippage: slippage,
        fees: fees,
        netCost: size * price + fees,
        fillProbability: Math.max(0, 100 - slippage),
        score: 0
      });
    }
    
    return results;
  }

  /**
   * Split large order across multiple platforms
   * @param {Object} params - Split parameters
   * @param {string} params.market - Market identifier
   * @param {string} params.side - Order side
   * @param {number} params.size - Total order size
   * @param {number} params.maxSlippage - Maximum allowed slippage
   */
  async splitOrder({ market, side, size, maxSlippage = 0.02 }) {
    const comparison = await this.compare({ market, side, size });
    
    // Filter platforms that can handle the order without exceeding slippage
    const viable = comparison.filter(p => p.slippage <= maxSlippage);
    
    if (viable.length === 0) {
      return { legs: [], totalSlippage: maxSlippage + 1, avgPrice: 0 };
    }
    
    // Distribute order based on liquidity and score
    const totalLiquidity = viable.reduce((sum, p) => sum + p.liquidity, 0);
    const legs = [];
    let remainingSize = size;
    
    for (let i = 0; i < viable.length - 1; i++) {
      const platform = viable[i];
      const allocation = Math.min(
        platform.liquidity * 0.8,
        remainingSize * (platform.liquidity / totalLiquidity)
      );
      legs.push({
        platform: platform.name,
        size: allocation,
        price: platform.price,
        estimatedSlippage: platform.slippage
      });
      remainingSize -= allocation;
    }
    
    // Last platform gets the remainder
    const last = viable[viable.length - 1];
    legs.push({
      platform: last.name,
      size: remainingSize,
      price: last.price,
      estimatedSlippage: last.slippage
    });
    
    const avgPrice = legs.reduce((sum, leg) => sum + leg.price, 0) / legs.length;
    const totalSlippage = legs.reduce((sum, leg) => sum + leg.estimatedSlippage, 0) / legs.length;
    
    return { legs, totalSlippage, avgPrice };
  }

  /**
   * Execute a routed order
   * @param {Object} route - Route object from findBestRoute
   */
  async execute(route) {
    // Simulate execution
    const actualPrice = route.expectedPrice * (1 + (Math.random() - 0.5) * route.expectedSlippage / 100);
    const actualSlippage = Math.abs((actualPrice - route.expectedPrice) / route.expectedPrice) * 100;
    
    return {
      orderId: `ord_${Date.now()}_${Math.random().toString(36).substr(2, 8)}`,
      platform: route.platform,
      fillPrice: actualPrice,
      actualSlippage: actualSlippage,
      fees: route.fees,
      executedAt: new Date().toISOString()
    };
  }

  /**
   * Analyze fee structures across platforms
   * @param {Object} params - Fee analysis parameters
   */
  async analyzeFees({ market, side, size }) {
    const results = [];
    
    for (const [platform, fees] of Object.entries(this.fees)) {
      const totalFee = (size * fees.taker) / 100;
      results.push({
        name: platform,
        makerFee: fees.maker,
        takerFee: fees.taker,
        totalFee: totalFee,
        hasRebate: fees.maker < 0,
        note: fees.note
      });
    }
    
    return results;
  }

  /**
   * Analyze liquidity across platforms
   * @param {Object} params - Liquidity analysis parameters
   */
  async analyzeLiquidity({ market, side }) {
    const results = [];
    
    for (const [platform, liquidity] of Object.entries(this.liquidityData)) {
      results.push({
        name: platform,
        bestBid: this.getBasePrice(market) * 0.99,
        bestAsk: this.getBasePrice(market) * 1.01,
        spread: liquidity.spread,
        depth1Pct: liquidity.depth1Pct,
        depth2Pct: liquidity.depth2Pct,
        note: this.fees[platform]?.note || ''
      });
    }
    
    return results;
  }

  // Helper methods
  getBasePrice(market) {
    // Mock price logic
    const prices = {
      'trump-2028': 0.52,
      'fed-rate-cut': 0.45,
      'btc-100k': 0.38
    };
    return prices[market] || 0.50;
  }

  getPlatformPrice(platform, basePrice, side) {
    // Mock platform-specific pricing (Polymarket usually has best price)
    const multipliers = {
      polymarket: 1.00,
      kalshi: 1.01,
      manifold: 1.02
    };
    return basePrice * (multipliers[platform] || 1.00);
  }

  createRoute(platformData, mode) {
    return {
      platform: platformData.name,
      mode: mode,
      expectedPrice: platformData.price,
      expectedSlippage: platformData.slippage,
      fees: platformData.fees,
      netCost: platformData.netCost,
      fillProbability: platformData.fillProbability,
      score: platformData.balancedScore || platformData.score
    };
  }
}

export default SmartRouter;
